File: runtime/learning-growth/python/generate_runtime_sbom.py
#!/usr/bin/env python3
from __future__ import annotations

import importlib.metadata
import json
import pathlib
import platform
import sys


def main() -> int:
    output = pathlib.Path(sys.argv[1] if len(sys.argv) > 1 else "learning-python-sbom.json")
    packages = sorted(
        ({"name": dist.metadata.get("Name", dist.name), "version": dist.version} for dist in importlib.metadata.distributions()),
        key=lambda row: (str(row["name"]).lower(), str(row["version"])),
    )
    payload = {
        "schema": "YANCE_LEARNING_RUNTIME_SBOM_V1",
        "runtime": "python",
        "python": platform.python_version(),
        "packages": packages,
    }
    output.write_text(json.dumps(payload, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: runtime/learning-growth/promptfoo/generate_runtime_sbom.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');

const output = path.resolve(process.argv[2] || 'learning-promptfoo-sbom.json');
const lock = JSON.parse(fs.readFileSync(path.resolve(__dirname, 'package-lock.json'), 'utf8'));
const packages = Object.entries(lock.packages || {}).filter(([key]) => key).map(([key, value]) => ({ path: key, version: value.version || '' })).sort((a, b) => a.path.localeCompare(b.path));
fs.writeFileSync(output, `${JSON.stringify({ schema: 'YANCE_LEARNING_RUNTIME_SBOM_V1', runtime: 'promptfoo', packages }, null, 2)}\n`, 'utf8');
